cmake: Add support for socs folder with Kconfig fragments/devicetree overlays

nordicjm opened the pull request:
Adds support for a new ``socs`` folder that can be placed in application folders and functions similar to the ``boards`` folder, but works for SoCs instead of boards

todo:

- [x] Doc update
- [x] Release notes update
- [x] Sample/test application showing usage

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1530,11 +1530,17 @@ endfunction()
 #                       [MERGE [REVERSE]]
 #   )
+#   zephyr_build_string(<out-variable>
+#                       SOC
```
`zephyr_build_string()` is self-contained so it should be provided with the `SOC` value, or the alternative, make the `BOARD` argument optional. If QUALIFIER is provided without BOARD, then we create the build string from just the qualifiers.

Also, using the `QUALIFIERS` together with SoC should be carefully considered, especially as we have a future enhancement in #69546:
> Support SiPs. Currently a SiP must be modeled as a SoC, although this is technically imprecise.
Multiple different SiPs could share same SoC, and thus could benefit from sharing SoC information.

A SiP field in beginning of board qualifiers could easily mess up the proposal here.

What I assume users really need is support for `qualifiers` overlay and conf fragments, not SoC overlays and conf fragments.

A qualifier approach would automatically support SiP qualifier in future.
Also the qualifier fragment is easier to reason why it should include board variant, than why a SoC fragment should consider board variants.

nordicjm replied:
The reason SOC cannot be provided is because CONFIG_SOC (or SOC) is not available until kconfig cmake code is loaded, which happens after this point, so the SOC has to be extracted from the board target name directly

tejlmand replied:
> The reason SOC cannot be provided is because CONFIG_SOC (or SOC) is not available until kconfig cmake code is loaded,

Which is exactly why I wrote:
> > or the alternative, make the BOARD argument optional. If QUALIFIER is provided without BOARD, then we create the build string from just the qualifiers.

So in following cases:

the call:
```
zephyr_build_string(<out> BOARD plank)
```
returns `plank`

the call:
```
zephyr_build_string(<out> BOARD plank BOARD_QUALIFIERS /foo/bar)
```
returns `plank_foo_bar`

and finally the call:
```
zephyr_build_string(<out> BOARD_QUALIFIERS /foo/bar)
```
returns `foo_bar`

The latter is what you're actually extending the function to support in this new feature.
But removing the SOC flag also removes any assumption as to whether the `BOARD_QUALIFIERS` contains a SoC, a SiP, or something else.

There is no need for the `SOC` flag in this function.

nordicjm replied:
This is not what it does though, the included help text shows what it returns:
```
# calling
#   zephyr_build_string(build_string SOC BOARD_QUALIFIERS /soc/bar/foo MERGE)
# will return a list of the following strings
# `soc_bar_foo;soc` in `build_string` parameter.
```

tejlmand replied:
> This is not what it does though, the included help text shows what it returns:

I can see what it returns, but please note that for the SOC comment, which I'm raising in the thread, then
1) the MERGE behavior has nothing to do with having a SOC or not parameter, thus irrelevant for the comment of removing the SOC option
2) The MERGE feature is another topic commented elsewhere.

tejlmand reviewed (commented):
Thanks for this proposal.

Although users has asked for SoC overlays and fragment, then I believe what they would actually want is qualifier based overlays and fragments.

That way we'll be naturally be supporting board SoC + board variants overlays and fragment in addition, as well as SiP overlays and fragments in future.

Can be achieved by making `BOARD` field optional and allow QUALIFIERS to be provided without BOARD argument.

nordicjm committed:
cmake: Add support for socs folder and Kconfig fragments

Adds support for a new ``socs`` folder that can be placed in
application folders and functions similar to the ``boards``
folder, but works for SoCs instead of boards

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
cmake: modules: extensions: Fix documentation for zephyr_build_string

Fixes a missed entry and a wrong named variable in the
documentation for this function

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
cmake: modules: configuration_files: Add support for soc overlays

Adds support for SoC overlay files which go in a ``socs`` folder
in application folders and functions similar to the ``boards``
folder, but works for SoCs instead of boards

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
tests: cmake: overlays: Add soc_folder_kconfig test

Adds a testcase for the new ``socs`` folder using Kconfig fragments

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
tests: cmake: overlays: Add soc_folder_overlay test

Adds a testcase for the new ``socs`` folder, with overlays

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: build: Add details on SoC overlays

Adds details about Kconfig fragments and devicetree overlay files
that will be included in a build if they are found in the
``socs`` folder

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.7: Add socs folder note

Adds release notes about the addition of a socs folder for
applications to store Kconfig fragments and devicetree overlays

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1583,17 +1594,45 @@ function(zephyr_build_string outvar)
 
-  if(DEFINED BUILD_STR_BOARD_QUALIFIERS AND NOT BUILD_STR_BOARD)
+  if(DEFINED BUILD_STR_BOARD_QUALIFIERS AND NOT BUILD_STR_BOARD AND NOT BUILD_STR_SOC)
     message(FATAL_ERROR
       "zephyr_build_string(${ARGV0} <list> BOARD_QUALIFIERS ${BUILD_STR_BOARD_QUALIFIERS} ...)"
-      " given without BOARD argument, please specify BOARD"
+      " given without BOARD or SOC argument, please specify BOARD or SOC"
+    )
+  endif()
```
All this code can be removed if we remove the need for `SOC` and instead allow `BOARD_QUALIFIERS` to be used without `BOARD`.

nordicjm replied:
Removed

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1583,17 +1594,45 @@ function(zephyr_build_string outvar)
+    )
+  endif()
+
+  if(BUILD_STR_SOC AND NOT BUILD_STR_MERGE)
+    message(FATAL_ERROR
+      "zephyr_build_string(${ARGV0} <list> SOC ...)"
+      " given without MERGE argument, this must be used with the SOC option"
```
MERGE should not be a requirement together with the SOC / BOARD_QUALIFIERS argument.

Keep zephyr_build_string generalized and let the caller decide if MERGE should be applied.
Note: see also other comment related to `MERGE`.

nordicjm replied:
Removed

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1571,8 +1577,13 @@ endfunction()
 # `alpha_bar_1_0_0;alpha_bar` in `short_build_string` parameter.
 #
+# calling
+#   zephyr_build_string(build_string SOC BOARD_QUALIFIERS /soc/bar/foo MERGE)
+# will return a list of the following strings
+# `soc_bar_foo;soc` in `build_string` parameter.
```
Not sure we should support MERGE from the start, based on experiences from #70548 and #71149 

In general we cannot expect that all overlays can apply cleanly, just think of the `rp2040.overlay` being applied for a `<board>/rp2040/w` build.

Let's re-consider a MERGE feature for qualifiers when addressing https://github.com/zephyrproject-rtos/zephyr/issues/69548.

nordicjm replied:
Removed, did not have rp2040 in mind when testing

tejlmand reviewed (changes requested):
Let's not introduce the MERGE feature yet, see comments for details.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -2572,7 +2617,7 @@ Please provide one of following: APPLICATION_ROOT, CONF_FILES")
     set(single_args APPLICATION_ROOT)
   elseif(${ARGV0} STREQUAL CONF_FILES)
-    set(options REQUIRED)
+    set(options SOC REQUIRED)
```
`QUALIFIERS` is a better option name, as what you want to pick up is overlays / conf files based upon the qualifiers and not only the SoC (SoC being part of the qualifiers, so it's also covered by that name)

```suggestion
    set(options QUALIFIERS REQUIRED)
```

nordicjm replied:
Updated

tejlmand commented on cmake/modules/configuration_files.cmake:
```diff
@@ -76,10 +76,19 @@ zephyr_boilerplate_watch(CONF_FILE)
 
 zephyr_get(DTC_OVERLAY_FILE SYSBUILD LOCAL)
 
+# Look for SoC overlay files, these should be prepended before board overlay files
+# but after app.overlay, depending upon which files have been found
```
this comment is a bit confusing, because it gives the reader the impression that a qualifier overlay will be used together with an `app.overlay`, which is not how the implementation works.

tejlmand replied:
Just had a second look.
Can see code comment is correct according to code (have code or my 👀 changed since earlier review ?) .

I'm not convinced that qualifier overlay should be used together with app.overlay.
If we have qualifier overlays behave identical to board overlay, then it works as an opt-out of an app.overlay, but with current proposal the app.overlay will still be included.

If following the board overlay approach, then a qualifier overlay can always do an
```
#include "app.overlay"
```
if it wants to include the `app.overlay`.

tejlmand reviewed (changes requested):
Thanks for this update.

Code wise, things looks correct according to expectations, however after last review and some more thinking of this, then I wonder if we should apply both qualifier and board overlay (similar for conf files) or have just one applied, that is from specific to general, like we do for other overlays and conf files.

Currently, if a `soc.overlay` is created, then it will be applied to both `plank/soc` and `plonk/soc`.
Ideally, the soc.overlay should apply cleanly regardless of board, but in some cases board  variants, like `plank/soc/ns` and `plonk/soc/ns` could have minor incompatible differences.

With current approach, `plonk` cannot opt-out of `socs/soc_ns.overlay`, as it will always be applied, even if a specific `boards/plonk_soc_ns.overlay` exists.

But if `socs/soc_ns.overlay` is only applied when `boards/plonk_soc_ns.overlay` doesn't exists, then potential edge cases can probably be avoided.

However, current proposal also have its pros.

Opening the discussion now, to minimize the risk of having to later open a discussion on this.

Besides this, rest looks good 👍

tejlmand commented on doc/build/dts/howtos.rst:
```diff
@@ -247,6 +247,8 @@ use as devicetree overlays:
 #. Otherwise, if :file:`<BOARD>.overlay` exists, it will be used, and the build
    system will stop looking for more files.
 #. Otherwise, if :file:`app.overlay` exists, it will be used.
+#. If the file :file:`socs/<SOC>_<BOARD_QUALIFIERS>.overlay` exists, it will be
+   included as an additional file
```
This mess up the docs, and makes it harder to follow the flow.

Because when reading step by step in existing doc, you can terminate the flow at multiple occasions:
https://github.com/zephyrproject-rtos/zephyr/blob/792fd57d96327dee3d7477479489ebd9fc42f907/doc/build/dts/howtos.rst?plain=1#L245-L246
and:
https://github.com/zephyrproject-rtos/zephyr/blob/792fd57d96327dee3d7477479489ebd9fc42f907/doc/build/dts/howtos.rst?plain=1#L247-L248

but those sentences are no longer true, because of this proposed `If the file` at the bottom of those steps.

Probably we should follow the existing overlay file structure and only used the qualifier overlay in addition to board overlay (and only board overlays from `boards/` folder), not app.overlay.

See also: https://github.com/zephyrproject-rtos/zephyr/pull/70418#discussion_r1580748053

nordicjm replied:
Have updated code and docs

tejlmand commented on cmake/modules/configuration_files.cmake:
```diff
@@ -91,9 +100,14 @@ if(NOT DEFINED DTC_OVERLAY_FILE)
   endif()
+endif()
 
+if(DEFINED soc_overlay_files)
+  list(APPEND DTC_OVERLAY_FILE ${soc_overlay_files})
```
This actually messes up when qualifier overlays are being applied, because here:
https://github.com/zephyrproject-rtos/zephyr/blob/af7b8400311cb903739554edc059d3592d6e168b/cmake/modules/configuration_files.cmake#L89
they are prepended to board specific overlays, but at this location:
https://github.com/zephyrproject-rtos/zephyr/blob/af7b8400311cb903739554edc059d3592d6e168b/cmake/modules/configuration_files.cmake#L106

they are appended to board overlays discovered here:
https://github.com/zephyrproject-rtos/zephyr/blob/af7b8400311cb903739554edc059d3592d6e168b/cmake/modules/configuration_files.cmake#L96

To keep both implementation and docs as simple as possible, let's follow how overlays are usually picked up, just with the smal adjustment that qualifier are used together with a board overlay from `boards/` folder, but no other overlays.
Similar to how revision overlays are used with board overlays from `board/` folder only.

nordicjm replied:
Have moved the soc overlay loading to the board overlay loading and prevented it loading app.overlay

tejlmand reviewed (changes requested):
It's too messy how qualifier overlays are picked up.
Both before and after board specific overlays, depending on where  those are located.
Also in addition to app.overlay (which none of the other overlay files does).

I prefer to start simple.
Uuse the qualifier overlays if found, and only together with board overlays from `boards/` if a valid overlay there is found.

fabiobaltieri reviewed (approved):
neat
